Use flock() instead of UUCP-style locking for serial devices

taw10 opened the pull request:
This replaces AcquireUUCPLockAndOpen() with AcquireLockAndOpen().  Depending on a compile-time option (--enable-uucp-locking), either flock() or UUCP-style lockfiles will be used as the underlying lock mechanism.  The default is to use flock().  The configure script will fall back on UUCP locking if flock() is not found.

It would also be possible to preserve the API by naming the top-level open/lock routine AcquireUUCPLockAndOpen, but then the routine name would not be an accurate description.  I've updated the plugins using these routines (only two of them, StageProfi and UsbPro).

ReleaseUUCPLock() is still present, becoming a no-op if flock() is being used.  With flock(), the lock is automatically released when the file descriptor is closed.

In practice, the TIOCEXCL lock (when available) is much more helpful, because it prevents any other process from opening the device and doesn't rely on all processes using the same locking scheme.  This lock is still used when flock() is available.

Fixes #1189, #1674 and many related problems on various platforms (e.g. running in a Toolbox container on Fedora Silverlight).

peternewman commented on common/io/Serial.cpp:
```diff
@@ -231,9 +268,11 @@ bool AcquireUUCPLockAndOpen(const std::string &path, int oflag, int *fd) {
+#endif // UUCP_LOCKING
 
 
 void ReleaseUUCPLock(const std::string &path) {
+#ifdef UUCP_LOCKING
```
I think it would be good to be consistent here too and create a ReleaseLock function, where the non UUCP path does nothing.

taw10 replied:
Done: added ReleaseSerialPortLock

peternewman commented on common/io/Serial.cpp:
```diff
@@ -27,6 +27,7 @@
 #include <signal.h>
 #include <string.h>
 #include <sys/stat.h>
+#include <sys/file.h>
```
Has sys/file.h existed forever, so we're not breaking UUCP builds which don't have it?

taw10 replied:
I'm not sure, I'll investigate.  However the header is checked by the configure script, and used without any #ifdefs in one other place (KarateLight).

peternewman replied:
Ah okay, obviously that plugin breaking wouldn't have quite so much impact, but the plugin has been around for a while and no-one has complained so we're probably good...

taw10 replied:
he flock call itself conforms to 4.4BSD, i.e. it's been around since 1997.  So I think we're pretty safe here.

I could add guards around the #include (ifdef HAVE_SYS_FILE_H), but if it's worked so far then I don't see any reason to add the extra complexity.   On inspection, I'm not actually sure why KarateLight includes this header at all.

peternewman replied:
Yeah very good point about the fact it's worked!

I'm not sure about KarateLight either; @cpresser ?

cpresser replied:
Unfortunately, I am not sure what to answer here. My C++ is quite rusty, I have not used it in years and sure have no idea which include contains which functionality.

I recall that I did include file locking of the serial-port-device in the karatelight code. A quick search turned up this piece of code here:
https://github.com/OpenLightingProject/ola/blob/master/plugins/karate/KarateLight.cpp#L121

Does that help?

taw10 replied:
The `#include <sys/file.h>` has been there since the first version of KarateDevice.cpp (f69a26b065adc853c3f7273d298c1589b1a5df68), but the `flock` call is in a different file (KarateLight.cpp).  I suspect the `flock` calls were originally in KarateDevice.cpp and got moved.

It does mean, however, that there's not only the #include but also a real use of flock() in OLA already without any tests!  Therefore we're probably pretty safe using it for all serial devices.  Nevertheless, we'll keep the tests and the conditionals :)

peternewman commented on common/io/Serial.cpp:
```diff
@@ -57,6 +58,7 @@ using std::string;
 
 namespace {
 
+#ifdef UUCP_LOCKING
```
I'm not sure of the benefit of not defining these if we're not in UUCP mode, it looks like they should work in all cases.

taw10 replied:
Ok, I removed the ifdefs, which reduces the noise a bit.

peternewman commented on common/io/Serial.cpp:
```diff
@@ -57,6 +58,7 @@ using std::string;
 namespace {
 
+#ifdef UUCP_LOCKING
 string GetLockFile(const string &path) {
```
This should probably be renamed though.

taw10 replied:
Done - renamed to GetUUCPLockFIle.

peternewman commented on common/io/Serial.cpp:
```diff
@@ -140,7 +143,41 @@ bool UIntToSpeedT(uint32_t value, speed_t *output) {
   return false;
 }
 
+#ifdef HAVE_FLOCK
+bool OpenAndFlock(const std::string &path, int oflag, int *fd) {
```
I think we've generally gone for having this the other way round and returning false, rather than this way where the build would break when it only impacts a small fraction of the functionality.

taw10 replied:
Ok, done.  Now it will try to open the device, but return false instead of locking it.

peternewman commented on common/io/Serial.cpp:
```diff
@@ -247,6 +286,16 @@ void ReleaseUUCPLock(const std::string &path) {
       OLA_INFO << "Released " << lock_file;
     }
   }
+#endif  /* else no action needed */
```
Please add an else and at least an OLA_DEBUG if not OLA_WARN, so the logs will show the function incorrectly getting called.

taw10 replied:
No longer relevant, I think.  ReleaseUUCPLock now unconditionally tries a UUCP-style unlock, so the error should be obvious if it gets called instead of the new ReleaseSerialPortLock.

peternewman replied:
Thanks, although please can we move the equivalent debug call to the new ReleaseSerialPortLock function instead please.

taw10 replied:
I've added an OLA_INFO message for the flock() path in ReleaseSerialPortLock.  Is that what you meant?  Or, did you mean to add a return value to ReleaseUUCPLock and check it in ReleaseSerialPortLock?

peternewman replied:
No that was what I meant, moving where the log message is generated.

peternewman commented on include/ola/io/Serial.h:
```diff
@@ -60,21 +60,27 @@ typedef enum {
+ * Depending on the compile-time configuration, this will use either flock()
+ * (the default) or UUCP locking.  See: ./configure --enable-uucp-locking.
+ *
  * This fails-fast, it we can't get the lock immediately, we'll return false.
  */
-bool AcquireUUCPLockAndOpen(const std::string &path, int oflag, int *fd);
+bool AcquireLockAndOpen(const std::string &path, int oflag, int *fd);
```
Please leave the other one documented, but deprecate it and @see the new function.

taw10 replied:
Ok, done.  The two pairs of routines (lock/unlock) now "@see" each other, and the two UUCP versions "@deprecated" to the corresponding new routines.

peternewman commented on common/io/Serial.cpp:
```diff
@@ -113,6 +115,7 @@ bool RemoveLockFile(const string &lock_file) {
   }
```
And this function ^ wants renaming too please.

taw10 replied:
Done - now it's RemoveUUCPLockFile.

peternewman reviewed (changes requested):
Hi @taw10 ,

Thanks for this!

There are some lint issues if you could fix them please:
https://app.travis-ci.com/github/OpenLightingProject/ola/jobs/558598505#L1483-L1487

If possible a test for this would be awesome! I.e. lock a file and try and open it again and confirm it fails etc. Which will then let people know during make test if their chosen locking hasn't worked.

taw10 commented:
Two more things still to do (add a test and check the situation with sys/file.h), but this should cover most of the comments.  The lint errors should be fixed (I may have introduced new errors, let's see).  I've also renamed the top-level lock/unlock routines, because "ReleaseLock" seemed too vague.  Now it's ReleaseSerialPortLock.

Note that in the flock() case, the unit test will really only test the TIOCEXCL locking, since that will prevent the open() call from succeeding before we try to flock().  Working round that would introduce some nasty complexity.  But, as long as one of the locking mechanisms works, it's OK.  With UUCP lockfiles, the check is done before we try to open the port itself.

peternewman commented on common/io/Serial.cpp:
```diff
@@ -157,11 +154,16 @@ bool OpenAndFlock(const std::string &path, int oflag, int *fd) {
     return false;
   }
+#else  // HAVE_FLOCK
+  close(*fd);
```
Probably chuck a log line in here saying it's tried to flock without flock, so someone doesn't get confused why it never locks if this branch is ever triggered.

taw10 replied:
Ok, done.  Now it will OLA_WARN "Tried to lock XXX, but flock() is unavailable.".

peternewman commented on common/io/Serial.cpp:
```diff
@@ -282,20 +282,26 @@ void ReleaseUUCPLock(const std::string &path) {
 }
 
-bool AcquireLockAndOpen(const std::string &path, int oflag, int *fd) {
+bool AcquireLockAndOpenSerialPort(const std::string &path, int oflag, int *fd) {
```
How come this only works for a serial port? Aside from the fact we've got it in Serial.cpp? Won't it work in any file?

taw10 replied:
Technically, flock() will work for any file of course.  UUCP wouldn't, though, because the lockfile name is made from the base name.  Semantically, the UUCP method is not really defined for anything other than a serial port.  So I don't think it makes sense to expand these routines to other things.

peternewman commented on include/ola/io/Serial.h:
```diff
@@ -59,6 +59,31 @@ typedef enum {
+ *
+ * This fails-fast, it we can't get the lock immediately, we'll return false.
+ *
+ * @deprecated Use AcquireLockAndOpenSerialPort() instead.
```
Please add a date to the deprecation.

taw10 replied:
Done!

peternewman commented on include/ola/io/Serial.h:
```diff
@@ -59,6 +59,31 @@ typedef enum {
+ *
+ * The lock is only removed if the PID matches.
+ *
+ * @deprecated Use ReleaseSerialPortLock() instead.
```
Again

taw10 replied:
Done!

peternewman commented on include/ola/io/Serial.h:
```diff
@@ -70,19 +95,24 @@ bool UIntToSpeedT(uint32_t value, speed_t *output);
  * @param path The path to unlock.
  *
- * The lock is only removed if the PID matches.
+ * If UUCP locking was used (see AcquireLockAndOpenSerialPort), the lockfile
```
I can't remember if this will auto-link, or whether you need to give the full namespace path for that.

taw10 replied:
I've checked, and added parentheses but not the whole path.  This makes it link correctly.

peternewman commented on plugins/usbpro/BaseUsbProWidget.cpp:
```diff
@@ -129,7 +129,7 @@ ola::io::ConnectedDescriptor *BaseUsbProWidget::OpenDevice(
   struct termios newtio;
   int fd;
-  if (!ola::io::AcquireLockAndOpen(path, O_RDWR | O_NONBLOCK | O_NOCTTY, &fd)) {
+  if (!ola::io::AcquireLockAndOpenSerialPort(path, O_RDWR | O_NONBLOCK | O_NOCTTY, &fd)) {
```
To fix the lint error:
```suggestion
  if (!ola::io::AcquireLockAndOpenSerialPort(path,
                                             O_RDWR | O_NONBLOCK | O_NOCTTY,
                                             &fd)) {
```

peternewman reviewed (changes requested):
A few more tweaks

peternewman commented:
> Two more things still to do (add a test and check the situation with sys/file.h), but this should cover most of the comments.

Great thanks.

> The lint errors should be fixed (I may have introduced new errors, let's see).

Only the one! :laughing: 

> I've also renamed the top-level lock/unlock routines, because "ReleaseLock" seemed too vague. Now it's ReleaseSerialPortLock.

As mentioned in the comment, won't this lock work on any file descriptor? Maybe the UUCP stuff is only implemented for devices?

> Note that in the flock() case, the unit test will really only test the TIOCEXCL locking, since that will prevent the open() call from succeeding before we try to flock(). Working round that would introduce some nasty complexity. But, as long as one of the locking mechanisms works, it's OK.

Exactly, I'm more interested in catching an issue during CI/build and around our logic than I am bothered about thoroughly testing this particular system feature. If we prove we can't open a file twice due to the lock then that's a good test as far as I'm concerned.

> With UUCP lockfiles, the check is done before we try to open the port itself.

Yep, understood.

taw10 commented:
Unit test is still to come (I have to get myself acquainted with the test system first).  One more question: would you prefer to have these commits squashed down to at least three (Codespell fix, unit test, everything else)?  The overall diff is very small and understandable.  Or do you prefer to have the complete history?

taw10 commented:
I've added a test, but it's not very satisfactory for a couple of reasons.  Let me know what you think!

The problem is that we don't have a real serial port to test with, and the TIOCEXCL lock won't work with a normal file.  To avoid creating a new file, I used the ChangeLog file for the test since it's always present at the top level of the project.  This feels bad to me, but the alternatives all seem less portable.  Maybe you or someone else has a suggestion.

To prevent the normal file from failing the test, I changed the two locking routines to not fail if `ioctl(fd, TIOCEXCL)` fails.  Perhaps this is an improvement, because we only need one of the three locking mechanisms to work (flock, UUCP or TIOCEXCL), but it also feels bad.

taw10 commented:
> To avoid creating a new file, I used the ChangeLog file for the test since it's always present at the top level of the project. This feels bad to me, but the alternatives all seem less portable. Maybe you or someone else has a suggestion.

This doesn't work with out-of-tree builds (nice that these work with OLA!), so now the test will create the file itself.  The file gets deleted afterwards.  If the file exists already for some reason (`serialLockTestFile` in the working directory), the test will fail.

peternewman commented on common/io/Serial.cpp:
```diff
@@ -299,6 +295,8 @@ bool AcquireLockAndOpenSerialPort(const std::string &path, int oflag, int *fd) {
 #ifdef UUCP_LOCKING
   ReleaseUUCPLock(path);
+#else   // UUCP_LOCKING
+  OLA_INFO << "No unlock necessary for " << path;
```
Can we change to one of these:
```suggestion
  OLA_INFO << "No unlock necessary with flock() for " << path;
```
or
```suggestion
  OLA_INFO << "No unlock necessary with for " << path << " as we're using flock()";
```

peternewman commented on common/io/Serial.cpp:
```diff
@@ -161,6 +161,7 @@ bool OpenAndFlock(const std::string &path, int oflag, int *fd) {
     return false;
   }
 #else  // HAVE_FLOCK
```
We don't normally bother commenting the else (and I personally think it makes it a bit more confusing as it's !HAVE_FLOCK).
```suggestion
#else
```

peternewman commented on common/io/Serial.cpp:
```diff
@@ -170,8 +171,6 @@ bool OpenAndFlock(const std::string &path, int oflag, int *fd) {
   // further opens.
   if (ioctl(*fd, TIOCEXCL) == -1) {
     OLA_WARN << "TIOCEXCL " << path << " failed: " << strerror(errno);
-    close(*fd);
```
Shouldn't we still close to tidy up?

taw10 replied:
It now closes (and unlocks, if necessary) on all exit paths.

peternewman commented on common/io/Serial.cpp:
```diff
@@ -170,8 +171,6 @@ bool OpenAndFlock(const std::string &path, int oflag, int *fd) {
   if (ioctl(*fd, TIOCEXCL) == -1) {
     OLA_WARN << "TIOCEXCL " << path << " failed: " << strerror(errno);
-    close(*fd);
-    return false;
```
Shouldn't it still return false if it's failed to TIOCEXCL?

taw10 replied:
It now closes (and unlocks, if necessary) on all exit paths.

peternewman commented on common/io/Serial.cpp:
```diff
@@ -263,9 +262,6 @@ bool AcquireUUCPLockAndOpen(const std::string &path, int oflag, int *fd) {
   // further opens.
   if (ioctl(*fd, TIOCEXCL) == -1) {
     OLA_WARN << "TIOCEXCL " << path << " failed: " << strerror(errno);
-    close(*fd);
```
Again we should close.

taw10 replied:
It now closes (and unlocks, if necessary) on all exit paths.

peternewman commented on common/io/Serial.cpp:
```diff
@@ -263,9 +262,6 @@ bool AcquireUUCPLockAndOpen(const std::string &path, int oflag, int *fd) {
   if (ioctl(*fd, TIOCEXCL) == -1) {
     OLA_WARN << "TIOCEXCL " << path << " failed: " << strerror(errno);
-    close(*fd);
-    RemoveUUCPLockFile(lock_file);
```
We should definitely try and tidy up shouldn't we so we don't leave a stale file?

peternewman commented on common/io/Serial.cpp:
```diff
@@ -263,9 +262,6 @@ bool AcquireUUCPLockAndOpen(const std::string &path, int oflag, int *fd) {
     OLA_WARN << "TIOCEXCL " << path << " failed: " << strerror(errno);
-    close(*fd);
-    RemoveUUCPLockFile(lock_file);
-    return false;
```
Again we should return false.

peternewman commented on common/io/SerialLockTester.cpp:
```diff
@@ -0,0 +1,73 @@
+void SerialLockTest::testLock() {
+  bool r1, r2;
+  int fd1, fd2, fd3;
+  const std::string path = "serialLockTestFile";
```
Given this could appear in system UUCP directories couldn't it?
```suggestion
  const std::string path = "olaSerialLockTestFile";
```

peternewman commented on common/io/SerialLockTester.cpp:
```diff
@@ -0,0 +1,73 @@
+
+  r1 = ola::io::AcquireLockAndOpenSerialPort(path, O_RDWR, &fd1);
+  OLA_ASSERT_TRUE(r1);
+
```
Perhaps check fd1 is valid/can be written to?

peternewman commented on include/ola/io/Serial.h:
```diff
@@ -68,7 +68,7 @@ bool UIntToSpeedT(uint32_t value, speed_t *output);
  * This fails-fast, it we can't get the lock immediately, we'll return false.
  *
- * @deprecated Use AcquireLockAndOpenSerialPort() instead.
+ * @deprecated Use AcquireLockAndOpenSerialPort() instead (19 Feb 2022).
```
```suggestion
 * @deprecated Use the more general AcquireLockAndOpenSerialPort() instead (19 Feb 2022).
```

peternewman reviewed (changes requested):
A few minor tweaks

peternewman commented:
> One more question: would you prefer to have these commits squashed down to at least three (Codespell fix, unit test, everything else)? The overall diff is very small and understandable.

If you see some of the commit history, you'll see I'm not too hot on that!

>  Or do you prefer to have the complete history?

Certainly while doing code reviews, I hate it when people force push as it breaks a lot of comment features within GitHub! I'm not really bothered either way and I think it occasionally helps when looking at blames etc.

peternewman commented on common/io/Makefile.mk:
```diff
@@ -47,7 +47,8 @@ test_programs += \
     common/io/StreamTester \
-    common/io/TimeoutManagerTester
+    common/io/TimeoutManagerTester \
+    common/io/SerialLockTester
```
Please add this new test to .gitignore.

taw10 replied:
It's actually already covered by these lines in .gitignore:
```
*Tester
*Tester.exe
```

peternewman commented:
> I've added a test, but it's not very satisfactory for a couple of reasons. Let me know what you think!

Great thanks.

> The problem is that we don't have a real serial port to test with, and the TIOCEXCL lock won't work with a normal file.

Is there any reason not to try locking a real serial port too if one exists? See the messaging ability at the bottom of this comment it could just fail saying maybe the port is in use. Especially if it tested ALL ports on a machine there's a good chance one would be available...

> To prevent the normal file from failing the test, I changed the two locking routines to not fail if `ioctl(fd, TIOCEXCL)` fails. Perhaps this is an improvement, because we only need one of the three locking mechanisms to work (flock, UUCP or TIOCEXCL), but it also feels bad.

Yeah it doesn't feel great to me.

Would it make sense to split each of the individual lock mechanism functions out, then you could have it like so (or whatever order it's currently in):
```
CheckFileExists
#ifdef flock
LockFlock
#else
LockUUCP
#endif
LockTIOCEXCL
```

So you could then test the constituent parts individually which would work around this issue.

> > To avoid creating a new file, I used the ChangeLog file for the test since it's always present at the top level of the project. This feels bad to me, but the alternatives all seem less portable. Maybe you or someone else has a suggestion.
> 
> This doesn't work with out-of-tree builds (nice that these work with OLA!), so now the test will create the file itself. The file gets deleted afterwards.

You can thanks Simon for the out-of-tree builds working. Although I've been scratching my head a bit getting a Python test to work with them...

> If the file exists already for some reason (`serialLockTestFile` in the working directory), the test will fail.

Presumably the test failing/bombing out once before deletion would cause all subsequent tests to fail afterwards?

See the best way to tell users that's the case: https://github.com/OpenLightingProject/ola/blob/c9add79a8d1c83ab209146006dae87906803b64f/ola/OlaClientWrapperTest.cpp#L52-L53

We could also possibly generate a random filename/suffix to test against to avoid some of that (although clutter up their drive if the test ever fails).

taw10 commented:
> > The problem is that we don't have a real serial port to test with, and the TIOCEXCL lock won't work with a normal file.
> 
> Is there any reason not to try locking a real serial port too if one exists? See the messaging ability at the bottom of this comment it could just fail saying maybe the port is in use. Especially if it tested ALL ports on a machine there's a good chance one would be available...

The idea of poking real hardware as part of `make check` worries me - we don't know what's on the other end of the serial port. Plus I'm not sure I want to write a load of serial port enumeration stuff (probably Linux-specific as well) just for this purpose.

> Would it make sense to split each of the individual lock mechanism functions out, then you could have it like so (or whatever order it's currently in):
> 
> ```
> CheckFileExists
> #ifdef flock
> LockFlock
> #else
> LockUUCP
> #endif
> LockTIOCEXCL
> ```
> 
> So you could then test the constituent parts individually which would work around this issue.

I'll implement this approach.  This way, we have a new locking mechanism (flock) and a test that it works, the UUCP stuff is still there as a fallback, and the TIOCEXCL is on top of either without a test.  This is already a big improvement over the old situation (legacy UUCP locking without a test).  The KarateLight experience (see above) suggests that the UUCP locking will _never_ be used, so the idea of testing UUCP might be purely theoretical anyway.

> Presumably the test failing/bombing out once before deletion would cause all subsequent tests to fail afterwards?

Unfortunately yes.  I'll add some cleanup so that the file gets removed in any case, and a message to make it clear that the file already exists.  However I want to avoid deleting a random user file in the event that my choice of filename wasn't "obscure" enough.

Update pending...

taw10 commented:
Ok, AcquireLockAndOpenSerialPort() now does this:
```
check port exists
if UUCP: acquire UUCP lock
open port
if FLOCK: apply flock()
if HAVE_IOCTL: apply TIOCEXCL
```
If any stage fails, it will bail out, closing the port and releasing UUCP lock if applicable.  So, all of the available locking mechanisms (potentially three of them, see below) have to succeed.  The bailout logic is a little more complicated than I'd like, because flock() happens after opening, whereas UUCP happens before.

AcquireUUCPLockAndOpen() is still available, although no-one should be using it.  It does this:
```
check port exists
acquire UUCP lock
open port
if HAVE_IOCTL: apply TIOCEXCL
```

I've split off `acquire UUCP lock` and `apply TIOCEXCL` as separate routines.

Each step prints at least an OLA_INFO on success, so you can see which locks are used.  If configured with `--enable-uucp-locking` and flock() is available, it will actually do both - there's no reason not to use flock() if the routine is available.  The information about KarateLight suggests that flock() is _always_ available, which would lead me to suggest completely removing the UUCP locking.  But perhaps there are systems still relying on it.

The test program (SerialLockTester) tests that the flock() call succeeds, but not that it prevents a second lock.  That would be a lot of trouble to test, because the second flock would have to be from a different process.  I consider it more a simple test that the routine really exists - it doesn't seem unreasonable to rely on it doing what the specification says it should!  It now uses a PID-based filename, to prevent one failure from causing subsequent ones.  UUCP locking and TIOCEXCL are not tested at all because of the difficulties noted above, but it's still overall an improvement on the previous situation.

peternewman commented on common/io/Serial.cpp:
```diff
@@ -249,21 +205,51 @@ bool AcquireUUCPLockAndOpen(const std::string &path, int oflag, int *fd) {
+    OLA_INFO << "Set TIOCEXCL on " << path;
+  }
+#else
+    OLA_INFO << "Not setting " << path << " - ioctl.h unavailable";
```
```suggestion
    OLA_INFO << "Not setting TIOCEXCL on " << path << " - ioctl.h unavailable";
```

peternewman commented on common/io/Serial.cpp:
```diff
@@ -249,21 +205,51 @@ bool AcquireUUCPLockAndOpen(const std::string &path, int oflag, int *fd) {
-  // further opens.
-  if (ioctl(*fd, TIOCEXCL) == -1) {
-    OLA_WARN << "TIOCEXCL " << path << " failed: " << strerror(errno);
+  if (!LockTIOCEXCL(*fd, path)) {
```
We may as well keep the comment:
```suggestion
  // As a final safety mechanism, use ioctl(TIOCEXCL) if available to prevent
  // further opens.
  if (!LockTIOCEXCL(*fd, path)) {
```

peternewman commented on common/io/Serial.cpp:
```diff
@@ -285,19 +271,55 @@ void ReleaseUUCPLock(const std::string &path) {
-#endif  // UUCP_LOCKING
+#else
+  OLA_INFO << "No unlock necessary for " << path << " (UUCP locking not used)";
+#endif
```
I think we do want the comment here to keep cpplint happy, just not on the else:
```suggestion
#endif  // UUCP_LOCKING
```

peternewman commented on common/io/SerialLockTester.cpp:
```diff
@@ -45,29 +46,38 @@ class SerialLockTest: public CppUnit::TestFixture {
 
-  OLA_ASSERT_FALSE(ola::io::FileExists(path));
+  std::stringstream str;
+  str << "serialLockTestFile." << our_pid;
```
Probably pedantry, but it would be nice if the file doesn't look like it's some sort of special file by ending in our PID...
```suggestion
  str << "serialLockTestFile." << our_pid << ".pid";
```

peternewman reviewed (commented):
Some minor tweaking

peternewman commented:
> The idea of poking real hardware as part of `make check` worries me - we don't know what's on the other end of the serial port.

I think we do already as part of our autostart tests TBH (but I can see your point)...

> Plus I'm not sure I want to write a load of serial port enumeration stuff (probably Linux-specific as well) just for this purpose.

I think that code should already exist as we already do essentially that in some of the plugins that will use this new locking!

> I'll implement this approach. This way, we have a new locking mechanism (flock) and a test that it works, the UUCP stuff is still there as a fallback, and the TIOCEXCL is on top of either without a test. This is already a big improvement over the old situation (legacy UUCP locking without a test). The KarateLight experience (see above) suggests that the UUCP locking will _never_ be used, so the idea of testing UUCP might be purely theoretical anyway.

Yeah fair point.

> > Presumably the test failing/bombing out once before deletion would cause all subsequent tests to fail afterwards?
> 
> Unfortunately yes. I'll add some cleanup so that the file gets removed in any case, and a message to make it clear that the file already exists. However I want to avoid deleting a random user file in the event that my choice of filename wasn't "obscure" enough.

Lovely thanks. That should be good enough, as long as the test fails with a non-cryptic message it shouldn't generate too many issues (judging by the auto-start stuff when it's already running as a service which obviously breaks more often).

peternewman commented on include/ola/io/Serial.h:
```diff
@@ -75,8 +78,41 @@ bool AcquireUUCPLockAndOpen(const std::string &path, int oflag, int *fd);
+ * @param[out] fd a pointer to the fd which is returned.
+ * @returns true if the open succeeded, false otherwise.
+ *
+ * Depending on the compile-time configuration, this will use either flock()
+ * (the default) or UUCP locking.  See: ./configure --enable-uucp-locking.
```
Isn't this now doing both/all?

peternewman commented on include/ola/io/Serial.h:
```diff
@@ -75,8 +78,41 @@ bool AcquireUUCPLockAndOpen(const std::string &path, int oflag, int *fd);
+ * If UUCP locking was used (see AcquireLockAndOpenSerialPort()), the lockfile
+ * will be removed (but only if the PID matches).
+ *
+ * Does nothing if flock() was used (see ./configure --enable-uucp-locking).
```
Again can't this do all/more now?

peternewman commented:
> Ok, AcquireLockAndOpenSerialPort() now does this:
> If any stage fails, it will bail out, closing the port and releasing UUCP lock if applicable. So, all of the available locking mechanisms (potentially three of them, see below) have to succeed. The bailout logic is a little more complicated than I'd like, because flock() happens after opening, whereas UUCP happens before.

Amazing!

> AcquireUUCPLockAndOpen() is still available, although no-one should be using it. It does this:

Cool.

> I've split off `acquire UUCP lock` and `apply TIOCEXCL` as separate routines.
> 
> Each step prints at least an OLA_INFO on success, so you can see which locks are used. If configured with `--enable-uucp-locking` and flock() is available, it will actually do both - there's no reason not to use flock() if the routine is available.

👍 

> The information about KarateLight suggests that flock() is _always_ available, which would lead me to suggest completely removing the UUCP locking. But perhaps there are systems still relying on it.

Yeah given some of the random architectures we've compiled on, they might just not have the KarateLight plugin included/compiled...

> The test program (SerialLockTester) tests that the flock() call succeeds, but not that it prevents a second lock. That would be a lot of trouble to test, because the second flock would have to be from a different process.

Ah yeah that would be quite a hassle! Although we could have a shell script wrapper which runs two copies if that's easy enough (like we do for some Python stuff), but as you say may not be worth the hassle.

> I consider it more a simple test that the routine really exists - it doesn't seem unreasonable to rely on it doing what the specification says it should! It now uses a PID-based filename, to prevent one failure from causing subsequent ones. UUCP locking and TIOCEXCL are not tested at all because of the difficulties noted above, but it's still overall an improvement on the previous situation.

Yeah sounds great!

taw10 commented:
Ok, some updates in addition to your suggested changes (which were all applied):
- Made sure that all the `#endif`s have comments
- Fixed the comments about the behaviour of `AcquireLockAndOpenSerialPort` and `ReleaseSerialPortLock`
- Made sure that `RemoveUUCPLockFile()` happens _after_ `close()` (as it was before)
- Added missing `ReleaseUUCPLock()` calls on the error paths of `AcquireLockAndOpenSerialPort`

taw10 commented:
The CI is failing, but the error doesn't make any sense to me (`ERROR: No matching distribution found for setuptools>=64`). Any suggestions?

DaAwesomeP commented:
@peternewman should this go into `0.10` or `master`?
